### PR TITLE
Add integration test for OSDs over PVCs

### PR DESCRIPTION
**Pull request by ashishranjan738**
Signed-off-by: Ashish Ranjan <aranjan@redhat.com>

This commit adds integration tests for osds provisioned over PVCs.

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

[test full]
[test ceph]

### CODE REVIEW

File: tests/integration/ceph_helm_test.go [hunk 1]
```diff
@@ -102,3 +102,61 @@ func (hs *HelmSuite) TestZFileStoreOnRookInstalledViaHelm() {
 func (hs *HelmSuite) TestObjectStoreOnRookInstalledViaHelm() {
 	runObjectE2ETestLite(hs.helper, hs.kh, hs.Suite, hs.namespace, "default", 3)
 }
+
+func TestCephHelmSuiteOnPVC(t *testing.T) {
```
Comment (travisn) [inline tests/integration/ceph_helm_test.go]: Instead of creating a new suite, I'd suggest modifying the existing helm suite. Two conditions need to be met in order to run the tests on PVCs:
- The k8s version is >= 1.13
- The `gp2` storage class exists in the cluster. (Our CI cluster might not have this storage class, we'll need to make sure it is there)

If those conditions are met, pass `true`, otherwise pass `false`. Then the helm test can run even if PVCs aren't available.

File: tests/framework/installer/ceph_manifests.go [hunk 2]
```diff
@@ -1675,6 +1676,52 @@ subjects:
...
+  - name: set1
+    count: 1
+    resources:
+    volumeClaimTemplates:
+    - metadata:
+        name: data
+      spec:
+        resources:
+          requests:
+            storage: 10Gi
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests.go]: for the tests let's request something small like 50Mi.

File: tests/framework/installer/ceph_manifests.go [hunk 3]
```diff
@@ -1675,6 +1676,52 @@ subjects:
...
+spec:
+cephVersion:
+  image: ` + settings.CephVersion.Image + `
+  allowUnsupported: ` + strconv.FormatBool(settings.CephVersion.AllowUnsupported) + `
+dataDirHostPath: ` + settings.DataDirHostPath + `
+network:
+  hostNetwork: false
+mon:
+  count: ` + strconv.Itoa(settings.Mons) + `
+  allowMultiplePerNode: true
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests.go]: In this test we should also use PVCs for mons.

File: tests/framework/installer/ceph_manifests.go [hunk 4]
```diff
@@ -1675,6 +1676,52 @@ subjects:
...
+dashboard:
+  enabled: true
+rbdMirroring:
+  workers: ` + strconv.Itoa(settings.RBDMirrorWorkers) + `
+metadataDevice:
+storage:
+  config:
+    storeType: "` + settings.StoreType + `"
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests.go]: We should always use bluestore for this test, so I'd suggest removing this line.

File: tests/framework/installer/ceph_installer.go [hunk 5]
```diff
@@ -183,7 +183,56 @@ func (h *CephInstaller) CreateK8sRookCluster(namespace, systemNamespace string,
 		NautilusVersion)
 }
 
-// CreateK8sRookCluster creates rook cluster via kubectl
+// CreateK8sRookClusterWithPVCs creates rook cluster with OSDs over pvc via kubectl
+func (h *CephInstaller) CreateK8sRookClusterWithPVCs(namespace, systemNamespace, storeType string,
```
Comment (travisn) [inline tests/framework/installer/ceph_installer.go]: Why do we need this method for installing on PVCs? I thought everything in the tests would be the same except the CephCluster CR spec would be different.

File: tests/framework/installer/ceph_installer.go [hunk 6]
```diff
@@ -314,31 +362,37 @@ func (h *CephInstaller) InstallRookOnK8sWithHostPathAndDevices(namespace, storeT
...
 
-	discovery, err := h.k8shelper.Clientset.AppsV1().DaemonSets(onamespace).Get("rook-discover", metav1.GetOptions{})
-	if startDiscovery {
-		assert.NoError(h.T(), err)
-		assert.NotNil(h.T(), discovery)
-	} else {
-		assert.Error(h.T(), err)
-		assert.True(h.T(), errors.IsNotFound(err))
+	if !usePVCs {
+		discovery, err := h.k8shelper.Clientset.AppsV1().DaemonSets(onamespace).Get("rook-discover", metav1.GetOptions{})
```
Comment (travisn) [inline tests/framework/installer/ceph_installer.go]: Whether the discovery service is running is actually an independent setting from whether PVCs are being used. I'd suggest we not change the discovery daemonset in this test. We can just leave it running.

**Comment by travisn** [comment]
@ashishranjan738 Instead of the helm test, the multi-cluster test would actually be a better candidate for the PVC tests. See [this explanation](https://github.com/rook/rook/pull/3495#issuecomment-546505409). Note that we would need to change the `multiClusterMinimalTestVersion` variable so the tests run on a later version of K8s in the CI. In PRs the suites are only run on a single version of K8s by default now.

File: tests/integration/ceph_helm_test.go (hunk 1 above)
Comment (ashishranjan738) [reply tests/integration/ceph_helm_test.go]: done

File: tests/framework/installer/ceph_manifests.go (hunk 4 above)
Comment (ashishranjan738) [reply tests/framework/installer/ceph_manifests.go]: done

File: tests/framework/installer/ceph_installer.go (hunk 5 above)
Comment (ashishranjan738) [reply tests/framework/installer/ceph_installer.go]: removed it

File: tests/framework/installer/ceph_installer.go (hunk 6 above)
Comment (ashishranjan738) [reply tests/framework/installer/ceph_installer.go]: cool removed it

File: tests/framework/installer/ceph_manifests.go (hunk 3 above)
Comment (ashishranjan738) [reply tests/framework/installer/ceph_manifests.go]: done

File: tests/integration/ceph_multi_cluster_test.go [hunk 7]
```diff
@@ -58,20 +58,22 @@ func TestCephMultiClusterDeploySuite(t *testing.T) {
...
 }
 
 // Deploy Multiple Rook clusters
 func (mrc *MultiClusterDeploySuite) SetupSuite() {
 
 	mrc.namespace1 = "mrc-n1"
 	mrc.namespace2 = "mrc-n2"
-
-	mrc.op, mrc.k8sh = NewMCTestOperations(mrc.T, mrc.namespace1, mrc.namespace2)
+	mrc.testPVCNamespace = "mrc-n3"
```
Comment (travisn) [inline tests/integration/ceph_multi_cluster_test.go]: Why do we need a third namespace? If we're going to test on PVCs, I thought we could just use namespace1 and namespace2 for those new OSDs. Let's just convert this whole test to a PVC test as long as conditions are met (1.13+ or other conditions). If the conditions are not met, we will not use PVCs. Everything else besides OSDs in the test can remain the same.

File: tests/integration/ceph_multi_cluster_test.go [hunk 8]
```diff
@@ -106,46 +115,67 @@ func (mrc *MultiClusterDeploySuite) TestInstallingMultipleRookClusters() {
...
 
 	kh, err := utils.CreateK8sHelper(t)
 	require.NoError(t(), err)
 	checkIfShouldRunForMinimalTestMatrix(t, kh, multiClusterMinimalTestVersion)
 
 	i := installer.NewCephInstaller(t, kh.Clientset, false, installer.VersionMaster, installer.NautilusVersion)
 
-	op := &MCTestOperations{i, kh, t, namespace1, namespace2, installer.SystemNamespace(namespace1)}
+	op := &MCTestOperations{i, kh, t, namespace1, namespace2, false, testPVCNamespace, installer.SystemNamespace(namespace1)}
+	p := kh.IsStorageClassPresent("gp2")
```
Comment (travisn) [inline tests/integration/ceph_multi_cluster_test.go]: In the CI log I see that `usePVCs=false`. We must need to do something to enable gp2 in the CI. This may be the trickiest part of enabling this test...

File: tests/integration/ceph_multi_cluster_test.go [hunk 9]
```diff
@@ -106,46 +115,67 @@ func (mrc *MultiClusterDeploySuite) TestInstallingMultipleRookClusters() {
...
+	if p == nil {
+		op.testPVC = true
```
Comment (travisn) [inline tests/integration/ceph_multi_cluster_test.go]: Is there a check for k8s 1.13+ somewhere?

File: tests/framework/installer/ceph_manifests.go [hunk 10]
```diff
@@ -1677,6 +1678,57 @@ subjects:
...
+    storageClassDeviceSets:
+    - name: set1
+      count: 1
+      volumeClaimTemplates:
+      - metadata:
+          name: data
+        spec:
+          resources:
+            requests:
+              storage: 10Gi
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests.go]: We can shrink this to 5Gi, right?

File: tests/framework/installer/ceph_manifests.go (hunk 2 above)
Comment (travisn) [reply tests/framework/installer/ceph_manifests.go]: We can shrink this to 1Gi?

File: tests/integration/ceph_multi_cluster_test.go (hunk 9 above)
Comment (ashishranjan738) [reply tests/integration/ceph_multi_cluster_test.go]: nope currently we don't have check

File: tests/integration/ceph_multi_cluster_test.go (hunk 7 above)
Comment (ashishranjan738) [reply tests/integration/ceph_multi_cluster_test.go]: I was thinking not to change the existing tests of hostPathDir instead of that we can test the osd over PVC on a separate namespace(mrc-n3). So what happens currently is if the `gp2` storageClass is present it will create a new namespace `mrc-n3` and and deploy a third ceph cluster(having osds deployed over PVCs) there and test it.

Comment (travisn) [reply tests/integration/ceph_multi_cluster_test.go]: Per a comment I linked previously, we eventually won't be able to have a multi-cluster test that runs on devices, or else we will need to worry about complications of the clusters using the same devices. It will be simplest if this multi-cluster test just uses PVCs. The other test suites will already cover the scenario of validating without PVCs, so no need for this suite to use both types of OSDs.

File: tests/framework/installer/ceph_manifests.go (hunk 2 above)
Comment (ashishranjan738) [reply tests/framework/installer/ceph_manifests.go]: done

File: tests/integration/ceph_multi_cluster_test.go (hunk 9 above)
Comment (ashishranjan738) [reply tests/integration/ceph_multi_cluster_test.go]: added it

File: tests/framework/installer/ceph_manifests.go (hunk 10 above)
Comment (ashishranjan738) [reply tests/framework/installer/ceph_manifests.go]: yup done

File: tests/integration/ceph_multi_cluster_test.go (hunk 8 above)
Comment (ashishranjan738) [reply tests/integration/ceph_multi_cluster_test.go]: don't have any clue how can we enable it. can we create it by ourselves?

**Comment by travisn** [comment]
@ashishranjan738 I wonder if we could base the PVC test on LVs after #4219 is merged. This would benefit both the general OSD on PVC code as well as the PVC on LV code. If we could remove the dependency from creating EBS volumes in the integration tests, it would be ideal since the new integration test could then also run outside AWS. @satoru-takeuchi do you have some ideas on how we could test OSDs on PVCs with the LVs?

**Comment by satoru-takeuchi** [comment]
@travisn One idea is to use TopoLVM, that the motivation of supporting OSD on LV-backed PVC, in the integration test. It can test OSD on PVC easily since TopoLVM is for the local volume and supports the dynamic provisioning feature. In addition, as you said, it can remove the dependency on EBS.

If we adopt this idea, merging this PR(#4151 ) first and merging a PR that use TopoLVM instead of EBS is nice. I can create the latter PR.

The cons of using TopoLVM in the integration test is that we can't test "OSD on raw-device-backed PVC" specific code. We should consider this case.

Since I'm not good at Rook's integration test, I would misunderstand something and I need to understand it more to make a better idea. I'm glad if you give me the document/code that is good to understand the detail of the integration test, For now, I can't find where the integration test uses EBS...

**Comment by satoru-takeuchi** [comment]
@travisn I forgot to tell the important information.

Why I propose TopoLVM is not because it's made by my project, but bacause it's the only CSI driver for local persistent volume with dynamic provisioning. In addition, lt will last long time since we will use it in our production environment. 

Of course, the best scenario is that we use the official local persistent volume. However, it doesn't support dynamic provisioner and it's difficult to support it near the future. Please take  a look at the current status of an issue for dynamic provisioner for local persistent volume. There is a design proposal but isn't any implementation yet.

https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner/issues/51

Since there is another CSI driver that is similar to TopoLVM, it's for PMEM as its name implies.

https://github.com/intel/pmem-csi

I consider the practical way to avoid to use remote storage like EBS is that we use TopoLVM for the time being and move to the official local persistent volume after it supports dynamic provisioner .

**Comment by travisn** [comment]
@satoru-takeuchi Great thoughts. Hopefully we can get both types of OSDs on PVCs into the integration tests so we can ensure there are no regressions in the future. Ideally maybe what we have in the multicluster test suite is:
- One namespace creates OSDs using EBS
- The other namespace creates OSDs with PVCs on LVs. TopoLVM sounds like it would work to test this since that's the purpose of this feature in rook.

The documentation for integration tests can be found [here](https://github.com/rook/rook/tree/master/tests), although it isn't very detailed. Maybe you could sync up with @ashishranjan738 to discuss the test.

File: pkg/operator/ceph/cluster/mon/mon.go [hunk 11]
```diff
@@ -424,7 +424,9 @@ func realScheduleMonitor(c *Cluster, mon *monConfig) (SchedulingResult, error) {
 
 	// setup affinity settings for pod scheduling
 	p := cephv1.GetMonPlacement(c.spec.Placement)
-	c.setPodPlacement(&d.Spec.Template.Spec, p, nil)
+	if !c.spec.Mon.AllowMultiplePerNode {
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/mon.go]: Please extract this and make it its own PR.

Comment (ashishranjan738) [reply pkg/operator/ceph/cluster/mon/mon.go]: yeah done https://github.com/rook/rook/pull/4741

**Comment by travisn** [comment]
@ashishranjan738 The multi-cluster test is passing. As far as I see from the logs it is based on the local PVCs that you created. Nice job to get this working! After #4742 merges, go ahead and rebase to fix the merge conflicts. The upgrade test failures should go away as well after the rebase.

**Comment by ashishranjan738** [comment]
> @ashishranjan738 The multi-cluster test is passing. As far as I see from the logs it is based on the local PVCs that you created. Nice job to get this working! After #4742 merges, go ahead and rebase to fix the merge conflicts. The upgrade test failures should go away as well after the rebase.

@travisn updated the PR. Really sry for my delay on this.
